pyomo/dae/flatten.py

```diff
@@ -234,47 +239,97 @@ def slice_component_along_sets(
-    of its parent component should be included in the stack.
-
-    `slice_` is the slice generated so far. Our goal here is to
-    yield extensions to `slice_` at this level of the hierarchy.
-
-    `sets` is a ComponentSet of Pyomo sets that should be sliced.
-
-    `ctype` is the type we are looking for.
+    Yields
+    ------
+    
+    Tuple of Sets and an IndexedComponent_slice or ComponentData
+        The sets indexing the returned component or slice. If the component
+        is indexed, an IndexedComponent_slice is returned. Otherwise, a
+        ComponentData is returned.
 
-    `index_map` is potentially a map from each set in `sets` to a 
-    "representative index" to use when descending into subblocks.
     """
     if type(slice_) is IndexedComponent_slice:
         context_slice = slice_.duplicate()
     else:
         context_slice = None
 
-    # Looks for components indexed by these sets immediately in our block
-    for c in b.component_objects(ctype, descend_into=False):
+    # If active argument is specified and does not match the block's
+    # active flag, we return immediately. This matches the behavior of
+    # component_objects. We only need this check as we may modify the
+    # active argument sent to component_objects if ctype is not an
+    # ActiveComponent type.
+    if active is not None and active != b.active:
+        return
+
+    # Define this class so we don't have to call issubclass again later.
+    check_active = issubclass(ctype, ActiveComponent) and (active != None)
+
+    # If active=False and ctype is not an ActiveComponent (e.g. it is Var)
+    # we will not generate any components. To prevent this, only pass the
+    # active argument if we are looking for active components.
+    c_active = active if check_active else None
+
+    # Looks for components indexed by specified sets immediately in our block.
+    for c in b.component_objects(ctype, descend_into=False, active=c_active):
         subsets = list(c.index_set().subsets())
         new_sets = [s for s in subsets if s in sets]
         # Extend our "index stack"
         sliced_sets = index_stack + new_sets
 
         # Extend our slice with this component
         for idx, new_slice in slice_component_along_sets(
-                c, sets, context_slice=context_slice, normalize=False
-                ):
-            yield sliced_sets, new_slice
+            c, sets, context_slice=context_slice, normalize=False
+        ):
+            # If we have to check activity, check data objects defined by
+            # slice. If any match, we yield the slice. This is done for
+            # compatibility with the behavior when slicing blocks, where
+            # we can only descend into a block that matches our active flag.
+            if not check_active or any(
+                data.active == c_active for data in new_slice.duplicate()
+            ):
+                yield sliced_sets, new_slice
```

Comment:
Calling `new_slice.duplicate()` causes problems for scalar components:

```
  File "C:\Users\[REDACTED]\Work\soec-control\soec_dynamic_flowsheet_helper.py", line 324, in <module>
    results = petsc.petsc_dae_by_time_element(
  File "c:\users\[REDACTED]\work\idaes-pse\idaes\core\solvers\petsc.py", line 489, in petsc_dae_by_time_element
    regular_cons, time_cons = flatten_dae_components(m, time, pyo.Constraint, active=True)
  File "c:\users\[REDACTED]\work\pyomo\pyomo\dae\flatten.py", line 556, in flatten_dae_components
    sets_list, comps_list = flatten_components_along_sets(
  File "c:\users\[REDACTED]\work\pyomo\pyomo\dae\flatten.py", line 483, in flatten_components_along_sets
    for index_sets, slice_ in generate_sliced_components(
  File "c:\users\[REDACTED]\work\pyomo\pyomo\dae\flatten.py", line 400, in generate_sliced_components
    for st, v in generate_sliced_components(
  File "c:\users\[REDACTED]\work\pyomo\pyomo\dae\flatten.py", line 400, in generate_sliced_components
    for st, v in generate_sliced_components(
  File "c:\users\[REDACTED]\work\pyomo\pyomo\dae\flatten.py", line 327, in generate_sliced_components
    data.active == c_active for data in new_slice.duplicate()
AttributeError: 'ScalarConstraint' object has no attribute 'duplicate'
python-BaseException
Backend TkAgg is interactive backend. Turning interactive mode on.
```

Reply:
Thanks! I'll add a test that covers this.

Reply:
Thanks for the fix. With the new update, it works for my application.